core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java

```diff
@@ -0,0 +1,73 @@
+package io.cucumber.core.logging;
+
+import io.cucumber.core.logging.LogRecordListener;
+import io.cucumber.core.logging.LoggerFactory;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Optional;
+
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@ExtendWith({ WithLogRecordListener.Extension.class })
+public @interface WithLogRecordListener {
+    class Extension implements BeforeEachCallback, AfterEachCallback, ParameterResolver {
+        private ExtensionContext.Store getContextStore(ExtensionContext context) {
+            Optional<Class<?>> testClass = context.getTestClass();
+            Object namespaceId = ExtensionContext.Namespace.GLOBAL;
+
+            // Use test class name as namespace for the context store
+            // If the test class is not available, fall back to the GLOBAL
+            // namespace
+            if (testClass.isPresent()) {
+                namespaceId = testClass.get().getSimpleName();
+            }
+
+            ExtensionContext.Namespace namespace = ExtensionContext.Namespace.create(namespaceId);
+            ExtensionContext.Store store = context.getStore(namespace);
+
+            return store;
```

Comment:
Try name spacing the store to the class of this extension and the test method. That should be sufficiently unique. And because you've already checked there is a test method in `supportsParameter` you won't have to check it again here. You can safely assume it to be present.